contracts/ERC20Safe.sol

```diff
@@ -205,19 +248,132 @@ contract ERC20Safe is Initializable, BridgeRole, Pausable {
         uint112 depositNonce;
-        (batchNonce, depositNonce) = _deposit_common(tokenAddress, amount, recipientAddress);
-        emit ERC20SCDeposit(batchNonce, depositNonce, callData);
+        (batchNonce, depositNonce) = _deposit_common(tokenAddress, amount, recipientAddress, callData);
     }
 
-
-    function _deposit_common(address tokenAddress, uint256 amount, bytes32 recipientAddress) internal returns (uint112 batchNonce, uint112 depositNonce) {
+    function _deposit_common(
+        address tokenAddress,
+        uint256 amount,
+        bytes32 recipientAddress,
+        bytes memory callData
+    ) internal returns (uint112 batchNonce, uint112 depositNonce) {
         require(whitelistedTokens[tokenAddress], "Unsupported token");
         require(amount >= tokenMinLimits[tokenAddress], "Tried to deposit an amount below the minimum specified limit");
         require(amount <= tokenMaxLimits[tokenAddress], "Tried to deposit an amount above the maximum specified limit");
 
+        _processDelayedTransactions(tokenAddress);
+
+        uint256 currentBlock = block.number;
+        uint256 bucketId = (currentBlock / blocksInBucket) % numBuckets;
+
+        _resetBucketIfNeeded(bucketId, tokenAddress, currentBlock);
+
+        if (amount >= singleTransactionThreshold[tokenAddress]) {
+            _addDelayedTransaction(tokenAddress, amount, recipientAddress, true, callData);
+            emit TransactionDelayed(msg.sender, tokenAddress, amount, recipientAddress, true);
+            return (0, 0);
+        } else {
+            uint256 totalAggregatedValue = _getTotalAggregatedValue(tokenAddress, currentBlock);
+            if (totalAggregatedValue + amount <= aggregateValueThreshold[tokenAddress]) {
+                aggregatedValue[bucketId][tokenAddress] += amount;
+                (batchNonce, depositNonce) = _processDeposit(tokenAddress, amount, recipientAddress, callData);
+                return (batchNonce, depositNonce);
+            } else {
+                _addDelayedTransaction(tokenAddress, amount, recipientAddress, false, callData);
+                emit TransactionDelayed(msg.sender, tokenAddress, amount, recipientAddress, false);
+                return (0, 0);
+            }
+        }
+    }
+
+    function _addDelayedTransaction(
+        address tokenAddress,
+        uint256 amount,
+        bytes32 recipientAddress,
+        bool isLarge,
+        bytes memory callData
+    ) internal {
+        DelayedTransaction memory dt = DelayedTransaction({
+            amount: amount,
+            tokenAddress: tokenAddress,
+            sender: msg.sender,
+            recipientAddress: recipientAddress,
+            blockAdded: block.number,
+            isLarge: isLarge,
+            callData: callData
+        });
+        delayedTransactions.push(dt);
+    }
+
+    function _processDelayedTransactions(address tokenAddress) internal {
+        uint256 i = 0;
+        while (i < delayedTransactions.length) {
```

Comment:
I guess this will consume a lot of gas. And the user y will actually pay for executing the deposit operations for user x, and so on. Thinking of a way to improve this and do the computation only when trying to assemble the current batch (I do not have a good idea now)

This method also has the disadvantage that, if a transaction is placed in a bucket, it won't be executed until a new deposit will be made. Which, if it will happen in 3 days, that will be the moment when it will actually be placed in a batch.